### PR TITLE
Improve SliceExt::binary_search performance

### PR DESCRIPTION
**Pull request by alkis**
Improve the performance of binary_search by reducing the number of unpredictable conditional branches in the loop. In addition improve the benchmarks to test performance in l1, l2 and l3 caches on sorted arrays with or without dups.

Before:

```
test slice::binary_search_l1                               ... bench:          48 ns/iter (+/- 1)
test slice::binary_search_l2                               ... bench:          63 ns/iter (+/- 0)
test slice::binary_search_l3                               ... bench:         152 ns/iter (+/- 12)
test slice::binary_search_l1_with_dups                     ... bench:          36 ns/iter (+/- 0)
test slice::binary_search_l2_with_dups                     ... bench:          64 ns/iter (+/- 1)
test slice::binary_search_l3_with_dups                     ... bench:         153 ns/iter (+/- 6)
```

After:

```
test slice::binary_search_l1                               ... bench:          15 ns/iter (+/- 0)
test slice::binary_search_l2                               ... bench:          23 ns/iter (+/- 0)
test slice::binary_search_l3                               ... bench:         100 ns/iter (+/- 17)
test slice::binary_search_l1_with_dups                     ... bench:          15 ns/iter (+/- 0)
test slice::binary_search_l2_with_dups                     ... bench:          23 ns/iter (+/- 0)
test slice::binary_search_l3_with_dups                     ... bench:          98 ns/iter (+/- 14)
```

### CODE REVIEW

File: src/libcore/benches/lib.rs
Comment (Mark-Simulacrum) [inline src/libcore/benches/lib.rs]: Why is this removed?

File: src/libcore/benches/slice.rs
Comment (Mark-Simulacrum) [inline src/libcore/benches/slice.rs]: `(0..999).collect::<Vec<_>>()` will be more efficient and somewhat easier to read (at least to me).

File: src/libcore/benches/slice.rs
Comment (Mark-Simulacrum) [inline src/libcore/benches/slice.rs]: Could we get a comment here as to what this number is intended to mean?

File: src/libcore/benches/lib.rs
Comment (alkis) [reply src/libcore/benches/lib.rs]: It doesn't exist.

File: src/libcore/benches/slice.rs
Comment (alkis) [reply src/libcore/benches/slice.rs]: Done.

File: src/libcore/benches/lib.rs
Comment (Mark-Simulacrum) [reply src/libcore/benches/lib.rs]: @alexcrichton How was this able to get past CI? Do we not run benches as part of libcore's tests?

File: src/libcore/benches/slice.rs
Comment (alkis) [reply src/libcore/benches/slice.rs]: It's a large (compared to 999) prime. Maybe I should use librand instead?

Comment (Mark-Simulacrum) [reply src/libcore/benches/slice.rs]: This seems fine to me -- and better than random numbers when dealing with benchmarks -- but I would like a comment.

Comment (alkis) [reply src/libcore/benches/slice.rs]: I made it into a proper LCG and linked to where I got the constants.

File: src/libcore/benches/lib.rs
Comment (cuviper) [reply src/libcore/benches/lib.rs]: I don't think we run benches in CI.  The file was moved in #44943.

**Comment by bluss** [comment]
I like it, trading an unpredictable Branch for a predictable one (bound check) is neat.

It doesn't exit when it finds the first equal element -- so does this PR change the result for some inputs?

**Comment by alkis** [comment]
The biggest benefit comes from trading all the unpredictable branches inside the loop with conditional moves. This is otherwise known as "branchless binary search" and has been shown to be faster than "traditional binary search" and as fast as linear search for small inputs. There is a recent paper that covers different layouts for comparison based searching: https://arxiv.org/abs/1509.05053. It covers a lot more than "branchless binary search", nevertheless it is a good read.

Answering your question: the PR should not change the results for any input. I added some extra test cases to increase confidence.

**Comment by arthurprs** [comment]
Cool! Although I think this needs to be tested further with other payloads, like (u64, u64), where the comparison isn't compiled into a single `cmov*`.

**Comment by alkis** [comment]
@arthurprs I don't think anything will change if (u64, u64) is used (or even strings for that matter). The `cmov` is used to select the new `base` after the comparison is done and its result is placed in the flags register: it is either `base` or `mid`. If the comparison operator has branches of its own that are not predictable there is not much we can do about it at this level. It has to be done by code external to `binary_search_by`.

**Comment by arthurprs** [comment]
Good point. I also ran a few benches just to confirm and it checks out.

**Comment by bluss** [comment]
@alkis there are no test cases that cover this — with duplicate elements. I've had time to check this at my computer now, and the new algorithm does produce different results for some inputs:

The example [(playground link)](https://play.rust-lang.org/?gist=a97e252cdad440c6b75525ef6d31d55e&version=nightly) is searching for 0 in a vector of zeros.

1. It's important to walk into breaking changes, however small, with open eyes
2. Now we know that it exists — can we find out if it matters? We've had bugs filed for breakage for less.

(Less important, but still interesting, is that the old algorithm will finish as quickly as possible with this input and the new algorithm it's a worst case.)

**Comment by alkis** [comment]
@bluss: thanks for testing this. I don't think these changes are material and very unlikely to result in breakages. Let me explain why:
* old: returns the first match. The match is implementation defined. Also it is not a well defined element: it is neither the lower bound nor the one before the upper bound. It is hard to make a case for using such match with any reasonable expectations.
* new: if a match exists it returns the lower bound. This is a stronger guarantee than the old implementation. This is an implementation detail but it has the risk of falling into [Hyrum's Law](http://www.hyrumslaw.com/) in the future.

In addition the two implementations will have different behavior when the list is not sorted given the partial order defined by the predicate. I also think that this is not important.

For performance it is true that the new algorithm always does `log2(size)` steps but the old does at most `log2(size)` steps. This is both good and bad. Bad because it can be slower in contrived situations and good because it has predictable performance.

I can add a few more benchmarks if you think they add value:
* string slices as keys
* with dups
* with different sizes: sizeof array fits in L1 cache, L2 cache and L3 cache

About possible breakages, if you have suggestions on how to investigate them before merging I would be glad to take a look.

**Comment by alkis** [comment]
I improved the benchmarks a bit. Please take another look.

**Comment by Mark-Simulacrum** [comment]
I think we should definitely get a crater run on this to at least note any test failures that would be caused by landing this PR. Is it possible to keep the old behavior without removing the performance gains this PR makes? I'm somewhat hesitant to change the behavior of this, and I'm not sure I entirely agree with all of your points about this being unlikely to hurt anyone in practice.

I agree that the new behavior on equal data is perhaps more elegant, but the old behavior (as I understand) is stable, if defined only by the algorithm used. Since it's stable, changing it now would lead me to believe that someone could depend on it -- even if they shouldn't be -- and we should be very hesitant to change it. Perhaps a survey of Rust code on GH that uses binary search could help here -- how much, if any, of it will change behavior given stretches of equal data? If we determine that probably none, then I'd be more inclined to make this change. If at least a couple different crates, then I'm pretty much against this.

With regards to unsorted data, I don't think there's any problem in changing behavior there -- a binary search on unsorted data isn't going to be reliable, at all, and this is something we explicitly document.

So, to summarize: I think that we should be very hesitant to land this without more data that this doesn't break people in practice (crater run and GH survey). It's "silent" breakage in that code continues to compile but changes in behavior, which is nearly the worst kind of breakage we can introduce.

r? @BurntSushi 

cc @rust-lang/libs -- how do we feel about the potential breakage here? Author's thoughts and the breakage are outlined in https://github.com/rust-lang/rust/pull/45333#issuecomment-337310949, mine are above.

**Comment by sfackler** [comment]
I'd be interested in seeing what a crater run turned up, but I'm not particularly worried about this change in behavior. We've never documented which of multiple equal values is arrived at, and it seems like you're in a bit of a weird place if you have that kind of data anyway.

**Comment by retep998** [comment]
Honestly, if it was going to pick a value from among equals, then I'd expect it to pick the lower bound. Because the current implementation picks at effectively random (even if it is deterministic), there's absolutely no way I'd be able to rely on *which* value it picked. The lower bound is still deterministic, so it won't break any code that wants deterministic behavior, but I have a *really* hard time imagining what sort of code can manage to rely on which specific element the current algorithm picks at random. Of course we should still do a crater run to be sure, but if we can't find any legitimate cases, then we should absolutely do this change.

**Comment by arthurprs** [comment]
I found myself wanting the lower bound of the equal subsequence recently. On the other hand I'd argue against guaranteeing this sort of behavior. Also, this discussion resembles the stable/unstable sort thing.

**Comment by kennytm** [comment]
The [documentation of `binary_search`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search) never guaranteed which index will be returned when there's an equal range. The new behavior is compatible with the existing definition.

> Looks up a series of four elements. The first is found, with a uniquely determined position; the second and third are not found; <ins>**the fourth could match any position in `[1, 4]`.**</ins>
>
> ```rust
> let s = [0, 1, 1, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55];
> 
> assert_eq!(s.binary_search(&13),  Ok(9));
> assert_eq!(s.binary_search(&4),   Err(7));
> assert_eq!(s.binary_search(&100), Err(13));
> let r = s.binary_search(&1);
> assert!(match r { Ok(1...4) => true, _ => false, });
> ```

**Comment by alkis** [comment]
@arthups I think not having lower_bound and upper_bound are holes in the std library. We can definitely add those. I can send a separate PR if there is consensus.

**Comment by frankmcsherry** [comment]
This appeals to me greatly. I can't recall if I complained to @bluss and that is why I got the ping, but (i) better performance is great, and (ii) being able to get to the lower bound is important for lots of applications. While this PR doesn't commit to that, it does mean that I could in principle start using `binary_search` which I couldn't do before (I mean, I could, but then I'd have to do geometric search backwards, and .. =/).

My understanding is that this *could* be slower if the comparison is very expensive, much like quicksort can be slower than mergesort if the comparison is expensive. Benchmarking on large randomly permuted `String` slices could tease that out (if each comparison is now an access to a random location in GBs of data).

@Mark-Simulacrum Rust has changed behavior a few times (for me wrt `sort`) and as the docs don't commit to any specific semantics other than "a" match, I'd be optimistic that this could land. I'm pro "check and see what breaks and try hard to avoid any", having just complained about that to Niko, but I'm also pro "actually commit to some semantics" and the first element is what I typically find needed.

**Comment by frankmcsherry** [comment]
```rust
if cmp == Equal { Ok(base) } else { Err(base + (cmp == Less) as usize) }
```

I don't know either way, but is `(cmp == Less) as usize` idiomatic? Is the compiler unable to turn a `match` statement into the right assembly? I didn't even realize that Rust commits to this being `0` or `1`.

**Comment by frankmcsherry** [comment]
In talking out the issue of intended semantics, what are the use cases where a collection may have multiple matching keys and returning "an arbitrary element" is acceptable? I'm familiar with "collection has distinct keys" in which things are unambiguous, and "collection has multiple keys, return the range please". I admit to never having seen an application that calls for what Rust currently offers; if someone could clue me in I would appreciate it!

**Comment by bluss** [comment]
I think this implementation is really neat but just using it in a new set of methods (for lower and upper bounds), leaving binary search unchanged sounds like the best solution. It makes the faster implementation available, it makes the useful lower bound algorithm available, and it avoids doing more steps than needed in `binary_search`. (Even if it looks like this extra work often has _“negative cost”_, we know there are cases where it has a cost.)

**Comment by alkis** [comment]
bluss@ it would extremely interesting to see cases that consistently get a regression out of this only because the old implementation bails out early. Also I don't think having a slow `binary_search` with a note saying "if you want performance use `lower_bound`" is a good choice for the standard library.

**Comment by alkis** [comment]
@frankmcsherry 

1. I don't see how this can be slower if the comparison is very expensive unless you are talking about getting lucky and landing on the correct element before you do N comparisons. @arthurprs benchmarked a tuple (which presumably is both more expensive and has an extra branch in the comparison) and found it having the same results as this implementation.
2. "branchless" benefits might go away if the comparison has branches in it. Maybe the CPU can detect dependent branches in this case and elide the cost of the second one. We can only know if we have a benchmark.
3. `(cmp == Less) as usize` not sure if this is idiomatic or not. I would be happy to switch to the idiomatic version if it has the same performance and is as succinct.
4. For sets with multiple matches the usual operation you want on it is `equal_range`. Perhaps we need to add that as well on top of `lower_bound` and `upper_bound`.

**Comment by arthurprs** [comment]
It's clear to me that lower_bound and upper_bound are desirable, deprecating (or not) binary_search. I fell that the name (self documenting/discoverable) is reason alone not to deprecate it though.

@alkis If you have "duplicates" and an expensive cmp function the new code might be slower.

**Comment by frankmcsherry** [comment]
@alkis 

1. It's not necessarily about getting very lucky. If you are doing a search by a key, rather than by element, you can find a key very early in the search and then spend the rest of your time finding the first element. The example from above was sorted strings and searching using just the first character to find a string starting with 'q'.

**Comment by bluss** [comment]
@alkis To be brief, you can reproduce the slower case using `let v = (0..size).map(|i| i / 32).collect::<Vec<_>>();` for the sorted sequence in the benchmark in the PR.

**Comment by alkis** [comment]
@frankmcsherry and @bluss see updated benchmarks:

Before:

```
test slice::binary_search_l1                               ... bench:          48 ns/iter (+/- 1)
test slice::binary_search_l2                               ... bench:          63 ns/iter (+/- 0)
test slice::binary_search_l3                               ... bench:         152 ns/iter (+/- 12)
test slice::binary_search_l1_with_dups                     ... bench:          36 ns/iter (+/- 0)
test slice::binary_search_l2_with_dups                     ... bench:          64 ns/iter (+/- 1)
test slice::binary_search_l3_with_dups                     ... bench:         153 ns/iter (+/- 6)
```

After:

```
test slice::binary_search_l1                               ... bench:          17 ns/iter (+/- 1)
test slice::binary_search_l2                               ... bench:          24 ns/iter (+/- 2)
test slice::binary_search_l3                               ... bench:         139 ns/iter (+/- 27)
test slice::binary_search_l1_with_dups                     ... bench:          17 ns/iter (+/- 1)
test slice::binary_search_l2_with_dups                     ... bench:          25 ns/iter (+/- 1)
test slice::binary_search_l3_with_dups                     ... bench:         137 ns/iter (+/- 21)
```

I don't see the regression.

Furthermore lets step back a bit. Do we expect no regressions? I do not think this is realistic. If we accept the fact there are going to be regressions we have to use [Amdhal's Law](https://en.wikipedia.org/wiki/Amdahl%27s_law) to assess the tradeoff. The performance increase on arrays that fit in L1 or L2 cache is about 2x. This is not trivial. So unless you think the regressions on the cases you think it regresses represent the majority of cases, I find it unwise to block this PR.

**Think of it in reverse. If the current code was the code in this PR, would be approve a change that changes it to the current code because it is faster on the contrived cases you mention?**

**Comment by alkis** [comment]
FWIW: I think the biggest risk of this change is that unit tests will break if they depend on the element/position returned by the current implementation of `binary_search`.

**Comment by Manishearth** [comment]
> **Think of it in reverse. If the current code was the code in this PR, would be approve a change that changes it to the current code because it is faster on the contrived cases you mention?**

I'm not sure if you noticed, but @frankmcsherry addressed this in an edit; "Edit: btw, I much prefer your code to what exists at the moment, which I don't use because it doesn't do what I need. I'm just trying to call out the things folks should worry about and be sure they are ok with.".


-----


Things seem to be getting a bit heated; it's worth mentioning that discussion like this is normal for changes to Rust; and it does not mean that folks wish to block your PR, only to aim for the best outcome.

To others: Let's continue to have clear technical discussion 😄 

Thanks for your contribution!

**Comment by alkis** [comment]
I am trying to run crater on windows but unfortunately I can't get the crater binary to compile: https://github.com/rust-lang-nursery/crater/issues/135#issuecomment-337741399. Any guidance on how to proceed?

**Comment by Manishearth** [comment]
Oh! Crater isn't something you run; it's something that someone will run for this PR.

Crater compiles every rustc crate that exists in the package repository (and also runs tests now); this is not something you want to run locally 😄 

I believe @Mark-Simulacrum is able to initiate crater runs (and will probably do so for this PR when they get the chance).

**Comment by alkis** [comment]
Fair enough. Even though it seems I should be able to do a mini crater run locally I will punt on it and work on something else until someone triggers larger crater run on this PR.

**Comment by Mark-Simulacrum** [comment]
I will start a Crater run on this as soon as it's fixed; added waiting-on-crater status.

**Comment by aidanhs** [comment]
@bors try
(prepare for crater run)

**Comment by alkis** [comment]
Slightly related: how do I go about adding `lower_bound`, `upper_bound` and `equal_range` to the standard library? PR/RFC?

**Comment by Manishearth** [comment]
Yeah, you'd have to RFC it. We're careful about additions to the stdlib.

**Comment by aidanhs** [comment]
This is behind a couple of other crater runs so will start being cratered in ~4 days.

**Comment by jonhoo** [comment]
I implemented the "best" version from the paper over at https://github.com/jonhoo/ordsearch (i.e., Eytzinger layout with prefetching), and am getting some pretty good speedups. Would probably be instructive to compare the two with different data sizes and lengths. Note that the crate requires nightly for prefetching since the [`prefetch`](https://github.com/aweinstock314/prefetch) crate requires nightly, and https://github.com/rust-lang/rfcs/pull/125 was closed.

**Comment by kennytm** [comment]
@jonhoo The existing `prefetch` intrinsic is not sufficient? Although it still requires nightly due to `#![feature(core_intrinsics)]`.

* https://doc.rust-lang.org/std/intrinsics/fn.prefetch_read_data.html

**Comment by jonhoo** [comment]
@kennytm ah, I hadn't seen that. My Google-fu must have been weak. That should be fine, although as you observe, it is still nightly-only.

**Comment by alkis** [comment]
@aidanhs is the crater backlog cleared? Can we run it now?

**Comment by aidanhs** [comment]
Crater run started - sorry for the delay, I've had a busy week!

**Comment by alkis** [comment]
Any results from the run? Is there a link to a dashboard I can inspect?

**Comment by aidanhs** [comment]
@alkis crater is currently a manual process, so it just depends on when someone gets a chance to see where things are up to and upload the results. It's done now though:

Crater results: http://cargobomb-reports.s3.amazonaws.com/pr-45333/index.html. 'Blacklisted' crates (spurious failures etc) can be found [here](https://github.com/rust-lang-nursery/crater/blob/master/blacklist.md). If you see any spurious failures not on the list, please make a PR against that file.

**Comment by Manishearth** [comment]
Please can we not ask contributors to do the work of analyzing crater output? That's our job. @alkis has done a lot of work here and shouldn't have further obligations to push this forward.

**Comment by Manishearth** [comment]
The failures for `kurnevsky.opai-rs.7c21e361514a6b21fa70b225b4e798cd15d76fac` and `zmq-pw-0.9.8` are concerning.

The latter is a `thread 'test_getset_plain_password' panicked at 'Timed out', /cargo-home/registry/src/github.com-1ecc6299db9ec823/timebomb-0.1.2/src/lib.rs:40:12`.

The former is a 

```
Oct 31 23:54:55.081 INFO blam! 	thread 'minimax_test::find_best_move' panicked at 'assertion failed: `(left == right)`
Oct 31 23:54:55.081 INFO blam!   left: `Some(37)`,
Oct 31 23:54:55.081 INFO blam!  right: `Some(36)`', src/minimax_test.rs:26:2
```

does someone have a downloadable rustc with this from the crater run? I could build my own but I'd rather not :)

**Comment by kennytm** [comment]
@Manishearth Currently, all compilers generated by `try` build can be downloaded (within 30 days) from https://s3-us-west-1.amazonaws.com/rust-lang-ci2/rustc-builds-try/7b76bc5df031c11773556a24ff2d294c4b7a4c5f/rust-nightly-x86_64-unknown-linux-gnu.tar.xz (the hash is the commit that is tried).

**Comment by Manishearth** [comment]
Ugh, that isn't rustup compatible; the stdlib is in a different folder :(

**Comment by Manishearth** [comment]
I'm a bit busy right now, can someone take point on checking the regressions with that compiler?

**Comment by alkis** [comment]
@kennytm is there is link to compiler at 7b76bc5df031c11773556a24ff2d294c4b7a4c5f for Mac OS?

**Comment by kennytm** [comment]
@alkis Nope, bors-try will only generate artifact for Linux. You need to setup a virtual machine or docker, or build the compiler yourself.

**Comment by kennytm** [comment]
@Manishearth @alkis 

The `kurnevsky.opai-rs` one is spurious. It is a flaky test which was recently fixed 4 days ago (see kurnevsky/opai-rs#5), but crater used an older commit  (kurnevsky/opai-rs@7c21e361514a6b21fa70b225b4e798cd15d76fac) right before the fix (kurnevsky/opai-rs@ac72d9382aa162b8aef94621f868dc1c4a1d80b2). You could disregard regression from it.

**Comment by alkis** [comment]
Good catch @kennytm! zmq failure is a timeout - perhaps its just flaky.

Looking at the bigger picture it seems that by large Hyrum's Law did not kick in yet and very few (if any) tests depend on the previous undocumented behavior of `binary_search`. So perhaps this PR is good to go?

**Comment by Manishearth** [comment]
Yeah, that timeout is likely spurious.

r? @rust-lang/libs

**Comment by alkis** [comment]
When merging please use the following message instead of the original.

---


Improve the performance of binary_search by reducing the number of unpredictable conditional branches in the loop.

Before:

```
test slice::binary_search_l1                               ... bench:          48 ns/iter (+/- 1)
test slice::binary_search_l2                               ... bench:          63 ns/iter (+/- 0)
test slice::binary_search_l3                               ... bench:         152 ns/iter (+/- 12)
test slice::binary_search_l1_with_dups                     ... bench:          36 ns/iter (+/- 0)
test slice::binary_search_l2_with_dups                     ... bench:          64 ns/iter (+/- 1)
test slice::binary_search_l3_with_dups                     ... bench:         153 ns/iter (+/- 6)
```

After:

```
test slice::binary_search_l1                               ... bench:          17 ns/iter (+/- 1)
test slice::binary_search_l2                               ... bench:          24 ns/iter (+/- 2)
test slice::binary_search_l3                               ... bench:         139 ns/iter (+/- 27)
test slice::binary_search_l1_with_dups                     ... bench:          17 ns/iter (+/- 1)
test slice::binary_search_l2_with_dups                     ... bench:          25 ns/iter (+/- 1)
test slice::binary_search_l3_with_dups                     ... bench:         137 ns/iter (+/- 21)
```

**Comment by Manishearth** [comment]
I just edited it in; merging is done by a bot so whatever is in your pull request message is what gets included.

**Comment by bluss** [comment]
@alkis You are free to update the PR title and the original PR message to better reflect the new contents. It's really good if you do! 

If you're can do it in git you can also rebase your changes and squash together commits to make the history neater. Feel free to do so after the code has been through a pass of reviews (I think it's fine now, right?)

File: src/libcore/slice/mod.rs
Comment (arthurprs) [inline src/libcore/slice/mod.rs]: We should probably use unsafe getter here to avoid the bounds check. The performance gain is minimal but there's no good reason to retain it now that we're going for full speed.

Same for `&s[base]` bellow

Comment (bluss) [reply src/libcore/slice/mod.rs]: I think the compiler optimizes them out -- at least when I checked a version of this PR's code earlier.

**Comment by alkis** [comment]
@bluss (sorry for the noobish questions, I am relatively new to git). So I `git rebase -i HEAD~N` where N is how many commits I want to squash and then `git push --force` on my fork of the repo? Isn't rewriting history that was already public a bad thing?

**Comment by Manishearth** [comment]
> Isn't rewriting history that was already public a bad thing?

For pull requests it's usually ok since it's not merged in anywhere.

It depends on the norms of a project.

File: src/libcore/slice/mod.rs
Comment (arthurprs) [reply src/libcore/slice/mod.rs]: I can see them in this test https://godbolt.org/g/jTCpAi but I may have screwed something.

**Comment by bluss** [comment]
@alkis Rebasing PRs here is fine. It's great if one doesn't rewrite all the time, so that it's easy to review additions changes but with this PR that phase is mostly over. You have a merge in the history of this PR too, and it would be great if that was removed, hopefully that comes out of your rebase.

I usually just look at what the base revision is (like the origin/master or what the origin/master was at some previous point and then I use `git rebase -i thatcommit` so that I don't have to count the commits.)

You don't need to rebase to the newest master or anything like that (unless there are conflicts). Basing it on any old revision that's still not conflicting is fine.

And another preemptive rebase tip, since contributors new and old here sometimes get submodule changes into their commits by mistake -- you don't want to commit submodule changes.

File: src/libcore/slice/mod.rs
Comment (alkis) [reply src/libcore/slice/mod.rs]: I swear I tried this before and it had no effect. Now it seems that it does so I am patching this in. I am squashing all commits in a bit.

**Comment by alkis** [comment]
@bluss Thanks a ton! I think I managed to get the rebase right. Please take a final look.

**Comment by alkis** [comment]
@aturon can you take a look? If you don't have time can you reassign to someone else?

**Comment by ishitatsuyuki** [comment]
@alkis is this really lower bound? I backported and it actually points to the last element of equal ones:
```
#[test]
    fn test_lower_bound() {
        let a = [1,1,1,1,1,1,1,1];
        assert_eq!(::binary_search(&a, &1), Ok(0)); // FAIL, result: Ok(7)
    }
```

**Comment by bluss** [comment]
It is important either way to add tests with duplicate elements.

**Comment by alkis** [comment]
@ishitatsuyuki I think you are correct. I am adding a second battery of tests that test implementation details.

**Comment by ishitatsuyuki** [comment]
I'm looking forward for having an exact lower bound guarantee; I'm currently copying and translating from SO every time I need them for algorithms.

For your reference, here's a collection of implementations on SO: https://stackoverflow.com/q/6443569/2484972

**Comment by bluss** [comment]
Side note for lower bounds. There is a somewhat ridiculous workaround for now @ishitatsuyuki to making binary search get the lower bound:

A comparison function that never returns `Equal`, so it has to find the correct boundary just lower than the cut off value . Works fine with duplicate elements. The ridiculous part is that one needs to unwrap the result since it's always `Err(index)`.

```rust
            data.binary_search_by(|x| if x >= cutoff {
                Ordering::Greater
            } else {
                Ordering::Less
            }).unwrap_err()
```

**Comment by alkis** [comment]
@bluss @ishitatsuyuki I tried different implementations that guarantee lower bound return on duplicate elements but they were all slower than the above. Do you have an implementation in mind that does this and gives the same performance?

**Comment by alkis** [comment]
Btw I have implementations of `lower_bound`, `upper_bound` and `equal_range` ready for https://github.com/rust-lang/rfcs/issues/2184. If we assume they will be part of stdlib then we do not have to make `binary_search` return the lower bound.

**Comment by ishitatsuyuki** [comment]
I also settled on a different comparison operator for bounds. The order shouldn't matter then.

**Comment by alkis** [comment]
For the impatient you can use `fast_binary_search` from https://crates.io/crates/ordslice.

**Comment by shepmaster** [comment]
Triage reminder for @bluss

**Comment by bluss** [comment]
I'll do a clarification and summary. Just correct me @alkis or others if I'm wrong.

- This PR improves performance of the binary search methods for slices
- The resulting index may change compared with the previous version when it comes to picking an element among a run of elements that are all equal
- The *lower bound* guarantee that was discussed in this thread, does not exist. (Important to underline so that there are no misunderstandings)
- This PR does not change what guarantees we are making for the binary_search methods
- There were no regressions found in the crater run

I personally think it will be no problem to introduce lower_bound, upper_bound etc methods to the std lib sooner or later, so there's no pressure to cram that functionality into the binary search methods.

**Comment by alkis** [comment]
@bluss this is great summary. Perhaps one thing missing is the reason for the lack of _lower_bound_ guarantee of `binary_search`: performance.

**Review by bluss — COMMENTED** [review]
Please add a test that exercises input data that is sorted, with runs of equal elements. This case has been an important case in the PR and we should cover it for the future. Maybe it is a good idea to leave a comment with those tests that we don't guarantee a particular output but we test the *current* behaviour of the algorithm.

File: src/libcore/slice/mod.rs
Comment (Manishearth) [inline src/libcore/slice/mod.rs]: It's pretty obvious, but it would be nice to have a tiny comment as to why this and the other `unsafe` are fine (should also explain that `mid` will never overflow).

**Review by Manishearth — APPROVED** [review]
With this small change and the test @bluss mentioned this should be ready to go.

r=me if necessary, though preferably @bluss should sign off

**Comment by arthurprs** [comment]
@alkis I'd like to point out that with this change using a binary search in the BTreeMap could be a win. Maybe you want to take a stab at that.

**Comment by ishitatsuyuki** [comment]
How? Binary trees are not contagious arrays, and the optimization techniques should be different.

**Comment by arthurprs** [comment]
They're **B**Tree's not BinaryTree's.

**Comment by alkis** [comment]
@ishitatsuyuki perhaps @arthurprs is talking about the array in each BTree node itself. Not sure what is happening there but it might be linear search instead of binary search.

File: src/libcore/slice/mod.rs
Comment (alkis) [reply src/libcore/slice/mod.rs]: Done.

File: src/libcore/tests/slice.rs
Comment (Manishearth) [inline src/libcore/tests/slice.rs]: The test is named pretty obviously, but please leave a comment stating that this is testing implementation-specific behavior of what to do in the case of equivalent elements; and that it is OK to break this but (as you've already mentioned) this should be accompanied with a crater run.

Comment (alkis) [reply src/libcore/tests/slice.rs]: Done.

**Comment by bluss** [comment]
@bors r+

Thanks a lot @alkis for this improvement!

**Comment by alkis** [comment]
Thanks for the thorough reviews!

**Pull request merged**
